<a href="https://colab.research.google.com/github/digitechit07/Python-Tutorial-with-Excercise/blob/main/Python_Math_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Performance Considerations**
The math module is implemented in C and is highly optimized for performance. However, keep these points in mind:

math functions work only with individual numbers, not sequences
For array operations, consider using NumPy instead
math.fsum() is more accurate but slower than sum() for floating-point addition
Use math.isclose() for floating-point comparisons instead of ==

In [1]:
import math as td

print('pi Constant - Pi = ', td.pi)
print('pi Constant - Degrees of Pi = ', td.degrees(td.pi))

print('\ne Constant - e = ', td.pi)
print('e Constant - Degrees of e = ', td.degrees(td.e))

print('\ntau Constant - tau = ', td.tau)
print('tau Constant - Degrees of tau = ', td.degrees(td.tau))

print('\ninf Constant - Positive Infinity = ', td.inf)
print('inf Constant - Negative Infinity = ', -td.inf)

print('\nNaN Constant - Not a Number = ', td.nan)


import math as mh

x = 10.98
y = 30.22
z = -40.95

print('FABS  - Absolute Value of z = ', mh.fabs(z))
print('FABS  - Absolute Value of -124.897 = ', mh.fabs(-124.897))

print('\ncopysign of x, z = ', mh.copysign(x, z))
print('copysign of z, x = ', mh.copysign(z, x))

print('\nCEIL  - Ceiling of x = ', mh.ceil(x))
print('CEIL  - Ceiling of y = ', mh.ceil(y))

print('\nFLOOR  - Floor of x = ', mh.floor(x))
print('FLOOR  - Floor of y = ', mh.floor(y))

print('\nFactorial of 3 = ', mh.factorial(3))
print('Factorial of 5 = ', mh.factorial(5))

import math as gm

print('FMOD - Mod of 2 and 3 = ', gm.fmod(2, 3))
print('FMOD - Mod of 225.55 and 5.5 = ', gm.fmod(222.55, 5.5))

print('\nFREXP - Mantissa and Exponent Value of 5 = ', gm.frexp(5))
print('FREXP - Mantissa and Exponent Value of -9 = ', gm.frexp(-9))

print('\nFSUM - Sum of Tuple Items = ', gm.fsum((10, 20, 30, 40)))
print('FSUM - Sum of List Items = ', gm.fsum([5, 22, 35, 9]))

print('\nGCD of two 10 and 2 = ', gm.gcd(10, 2))
print('GCD of two 100 and 15 = ', gm.gcd(100, 15))


import math as at

print('ROUND - Rounded Number 100.98763 = ', round(100.9876, 2))
print('ROUND - Rounded Number 125.932832 = ', round(125.932832, 3))

print('\nLDEXP - LDEXP (FREXP inverse) Number of 4, 5 = ', at.ldexp(4, 5))
print('LDEXP - LDEXP (FREXP inverse) Number of -9, 2 = ', at.ldexp(-9, 2))

print('\nMODF - Modf (Divided 1 to 2) Number of 100 = ', at.modf(100))
print('MODF - Modf (Divided 1 to 2) Number of 120.98 = ', at.modf(120.98))

print('\nTRUNC - Truncated Number 100.98763 = ', at.trunc(100.9876))
print('ROUND - Truncated Number 125.932832 = ', at.trunc(-125.932832))

print('\nRemainder of 29 and 5 = ', at.remainder(20, 5))
print('Remainder of 10 and 3 = ', at.remainder(10, 3))

import math as th

print('exp of 5 = ', th.exp(5))
print('exp of -3 = ', th.exp(-3))

print('\nexpm1 of 8 = ', th.expm1(8))
print('expm1 of -5 = ', th.expm1(-5))

print('\nLOG  - logarithmic of 5 = ', th.log(5))
print('LOG  - logarithmic of 100 Base 2 = ', th.log(100, 2))

print('\nLOG2  - logarithmic of 120 Base 2 = ', th.log2(120))

print('\nLOG10  - logarithmic of 150 Base 10 = ', th.log2(150))

print('\nPOW  - 2 Power 3  = ', th.pow(2, 3))
print('POW  - 5 Power 4  = ', th.pow(5, 4))

print('\nSQRT  - Square Root of 25 = ', th.sqrt(25))
print('SQRT  - Square Root of 19 = ', th.sqrt(19))

import math as mt

print('COS  - Cosine of 10 = ', mt.cos(10))
print('COS  - Cosine of -15 = ', mt.cos(-15))

print('\nSIN  - Sine of 3 = ', mt.sin(3))
print('SIN  - Sine of -5 = ', mt.sin(-5))

print('\nTAN  - Tangent of 9 = ', mt.tan(9))
print('TAN  - Tangent of -3 = ', mt.tan(-3))

print('\nACOS  - Arc Cosine of 1 = ', mt.acos(1))
print('ACOS  - Arc Cosine of -0.78 = ', mt.acos(-0.78))

print('\nASIN  - Arc Sine of 1 = ', mt.asin(1))
print('ASIN  - Arc Sine of -2 = ', mt.asin(-0.42))

print('\nATAN  - Arc Tangent of 0.72 = ', mt.atan(0.72))
print('ATAN  - Arc Tangent of -2.71 = ', mt.atan(-2.71))

print('\nATAN2  - Tangent of 2, 5 = ', mt.atan2(2, 5))

print('\nHYPOT  - Hypot Value of 2, 3 = ', mt.hypot(2, 3))


import math as ma

print('COSH  - Hyperbolic Cosine of 2 = ', ma.cosh(2))
print('COSH  - Hyperbolic Cosine of -1 = ', ma.cosh(-1))

print('\nSINH  - Hyperbolic Sine of 3 = ', ma.sinh(3))
print('SINH  - Hyperbolic Sine of -5 = ', ma.sinh(-5))

print('\nTANH  - Hyperbolic Tangent of 1 = ', ma.tanh(1))
print('TANH  - Hyperbolic Tangent of -3 = ', ma.tanh(-3))

print('\nACOSH  - Hyperbolic Arc Cosine of 10 = ', ma.acosh(10))
print('ACOSH  - Hyperbolic Arc Cosine of 30.78 = ', ma.acosh(30.78))

print('\nASINH  - Hyperbolic Arc Sine of 15 = ', ma.asinh(15))
print('ASINH  - Hyperbolic Arc Sine of -25 = ', ma.asinh(-25))

print('\nATANH  - Hyperbolic Arc Tangent of 0.57 = ', ma.atanh(0.57))
print('ATANH  - Hyperbolic Arc Tangent of -0.71 = ', ma.atanh(-0.71))

import math as gd

print('DEGREES Function - Degrees Value of 6 = ', gd.degrees(5))
print('DEGREES Function - Degrees Value of 12 = ', gd.degrees(12))

print('\nRADIANS Function - Radians Value of 350 = ', gd.radians(350))
print('\nRADIANS Function - Radians Value of 680 = ', gd.radians(680))

print('\nGAMMA Function - Gamma Value of 8 = ', gd.gamma(8))
print('LGAMMA Function - LGamma Value of 9 = ', gd.lgamma(9))

# This function adds two numbers
def add(x, y):
    return x + y

# This function subtracts two numbers
def subtract(x, y):
    return x - y

# This function multiplies two numbers
def multiply(x, y):
    return x * y

# This function divides two numbers
def divide(x, y):
    return x / y


print("Select operation.")
print("1.Add")
print("2.Subtract")
print("3.Multiply")
print("4.Divide")

while True:
    # take input from the user
    choice = input("Enter choice(1/2/3/4): ")

    # check if choice is one of the four options
    if choice in ('1', '2', '3', '4'):
        try:
            num1 = float(input("Enter first number: "))
            num2 = float(input("Enter second number: "))
        except ValueError:
            print("Invalid input. Please enter a number.")
            continue

        if choice == '1':
            print(num1, "+", num2, "=", add(num1, num2))

        elif choice == '2':
            print(num1, "-", num2, "=", subtract(num1, num2))

        elif choice == '3':
            print(num1, "*", num2, "=", multiply(num1, num2))

        elif choice == '4':
            print(num1, "/", num2, "=", divide(num1, num2))

        # check if user wants another calculation
        # break the while loop if answer is no
        next_calculation = input("Let's do next calculation? (yes/no): ")
        if next_calculation == "no":
          break
    else:
        print("Invalid Input")




pi Constant - Pi =  3.141592653589793
pi Constant - Degrees of Pi =  180.0

e Constant - e =  3.141592653589793
e Constant - Degrees of e =  155.74607629780772

tau Constant - tau =  6.283185307179586
tau Constant - Degrees of tau =  360.0

inf Constant - Positive Infinity =  inf
inf Constant - Negative Infinity =  -inf

NaN Constant - Not a Number =  nan
FABS  - Absolute Value of z =  40.95
FABS  - Absolute Value of -124.897 =  124.897

copysign of x, z =  -10.98
copysign of z, x =  40.95

CEIL  - Ceiling of x =  11
CEIL  - Ceiling of y =  31

FLOOR  - Floor of x =  10
FLOOR  - Floor of y =  30

Factorial of 3 =  6
Factorial of 5 =  120
FMOD - Mod of 2 and 3 =  2.0
FMOD - Mod of 225.55 and 5.5 =  2.5500000000000114

FREXP - Mantissa and Exponent Value of 5 =  (0.625, 3)
FREXP - Mantissa and Exponent Value of -9 =  (-0.5625, 4)

FSUM - Sum of Tuple Items =  100.0
FSUM - Sum of List Items =  71.0

GCD of two 10 and 2 =  2
GCD of two 100 and 15 =  5
ROUND - Rounded Number 100.98763 =  10

# **Key takeaways for using the math module effectively:**

Import only the functions you need for better code readability
Be mindful of angle units (radians vs degrees) in trigonometric functions
Use appropriate functions for floating-point comparisons and summation
Handle edge cases and domain errors appropriately
Consider NumPy for array-based mathematical operations
Leverage the module’s accuracy and performance optimizations

In [3]:
import time

big_list = list(range(1000000))
big_set = set(big_list)
start = time.time()
print(999999 in big_list)
print(f"List lookup: {time.time() - start:.6f}s")

start = time.time()
print(999999 in big_set)
print(f"Set lookup: {time.time() - start:.6f}s")


numbers = list(range(1000000))
def modify_list(lst):
    lst[0] = 999
    return lst
start = time.time()
result = modify_list(numbers)
print(f"In-place: {time.time() - start:.4f}s")

def copy_list(lst):
    new_lst = lst.copy()
    new_lst[0] = 999
    return new_lst
start = time.time()
result = copy_list(numbers)
print(f"Copy: {time.time() - start:.4f}s")

class Point:
    __slots__ = ('x', 'y')
    def __init__(self, x, y):
        self.x = x
        self.y = y
start = time.time()
points = [Point(i, i+1) for i in range(1000000)]
print(f"With slots: {time.time() - start:.4f}s")

import math
numbers = list(range(10000000))
start = time.time()
roots = [math.sqrt(n) for n in numbers]
print(f"Math sqrt: {time.time() - start:.4f}s")

start = time.time()
roots = [n ** 0.5 for n in numbers]
print(f"Operator: {time.time() - start:.4f}s")

start = time.time()
result = [0] * 1000000
for i in range(1000000):
    result[i] = i
print(f"Pre-allocated: {time.time() - start:.4f}s")

start = time.time()
result = []
for i in range(1000000):
    result.append(i)
print(f"Dynamic: {time.time() - start:.4f}s")

numbers = list(range(10000000))
start = time.time()
total = 0
for i in numbers:
    if i % 2 != 0:
        total += i // 2
    else:
        total += i
print(f"Conditional: {time.time() - start:.4f}s")

start = time.time()
total = 0
for i in numbers:
    try:
        total += i / (i % 2)
    except ZeroDivisionError:
        total += i
print(f"Exception: {time.time() - start:.4f}s")

def outer():
    def add_pair(a, b):
        return a + b
    result = 0
    for i in range(10000000):
        result = add_pair(result, i)
    return result
start = time.time()
result = outer()
print(f"Local function: {time.time() - start:.4f}s")

def add_pair(a, b):
    return a + b
start = time.time()
result = 0
for i in range(10000000):
    result = add_pair(result, i)
print(f"Global function: {time.time() - start:.4f}s")

from itertools import product
items = [1, 2, 3] * 10
start = time.time()
result = list(product(items, repeat=2))
print(f"Itertools: {time.time() - start:.4f}s")

start = time.time()
result = []
for x in items:
    for y in items:
        result.append((x, y))
print(f"Loops: {time.time() - start:.4f}s")

import bisect
numbers = sorted(list(range(0, 1000000, 2)))
start = time.time()
bisect.insort(numbers, 75432)
print(f"Bisect: {time.time() - start:.4f}s")

start = time.time()
for i, num in enumerate(numbers):
    if num > 75432:
        numbers.insert(i, 75432)
        break
print(f"Loop: {time.time() - start:.4f}s")

def expensive_operation():
    time.sleep(0.001)
    return 42
start = time.time()
cached_value = expensive_operation()
result = 0
for i in range(1000):
    result += cached_value
print(f"Cached: {time.time() - start:.4f}s")

start = time.time()
result = 0
for i in range(1000):
    result += expensive_operation()
print(f"Repeated: {time.time() - start:.4f}s")

from statistics import median
from math import isnan
from itertools import filterfalse

data = [20.7, float('NaN'),19.2, 18.3, float('NaN'), 14.4]
sorted(data)  # This has surprising behavior

median(data)  # This result is unexpected


sum(map(isnan, data))    # Number of missing values

clean = list(filterfalse(isnan, data))  # Strip NaN values
clean

sorted(clean)  # Sorting now works as expected

median(clean)       # This result is now well defined

from statistics import mean
mean([1, 2, 3, 4, 4])

mean([-1.0, 2.5, 3.25, 5.75])


from fractions import Fraction as F
mean([F(3, 7), F(1, 21), F(5, 3), F(1, 3)])


from decimal import Decimal as D
mean([D("0.5"), D("0.75"), D("0.625"), D("0.375")])
'''
data = [-2.1, -1.3, -0.4, 1.9, 5.1, 6.2]
rand = kde_random(data, h=1.5, seed=8675309)
new_selections = [rand() for i in range(10)]
[round(x, 1) for x in new_selections]


import statistics
from decimal import Decimal as D
pvariance([D("27.5"), D("30.25"), D("30.25"), D("34.5"), D("41.75")])


from fractions import Fraction as F
pvariance([F(1, 4), F(5, 4), F(1, 2)])

from decimal import Decimal as D
variance([D("27.5"), D("30.25"), D("30.25"), D("34.5"), D("41.75")])


from fractions import Fraction as F
variance([F(1, 6), F(1, 2), F(5, 3)])


data = [105, 129, 87, 86, 111, 111, 89, 81, 108, 92, 110,
        100, 75, 105, 103, 109, 76, 119, 99, 91, 103, 129,
        106, 101, 84, 111, 74, 87, 86, 103, 103, 106, 86,
        111, 75, 87, 102, 121, 111, 88, 89, 101, 106, 95,
        103, 107, 101, 81, 109, 104]
[round(q, 1) for q in quantiles(data, n=10)]

x = [1, 2, 3, 4, 5, 6, 7, 8, 9]
y = [1, 2, 3, 1, 2, 3, 1, 2, 3]
covariance(x, y)

z = [9, 8, 7, 6, 5, 4, 3, 2, 1]
covariance(x, z)

covariance(z, x)
'''

True
List lookup: 0.011541s
True
Set lookup: 0.000092s
In-place: 0.0001s
Copy: 0.0048s
With slots: 1.8380s
Math sqrt: 1.9008s
Operator: 1.2152s
Pre-allocated: 0.1133s
Dynamic: 0.1003s
Conditional: 1.6772s
Exception: 3.0192s
Local function: 1.0596s
Global function: 1.7677s
Itertools: 0.1660s
Loops: 0.0002s
Bisect: 0.0009s
Loop: 0.0089s
Cached: 0.0018s
Repeated: 1.0945s


'\ndata = [-2.1, -1.3, -0.4, 1.9, 5.1, 6.2]\nrand = kde_random(data, h=1.5, seed=8675309)\nnew_selections = [rand() for i in range(10)]\n[round(x, 1) for x in new_selections]\n\n\nimport statistics\nfrom decimal import Decimal as D\npvariance([D("27.5"), D("30.25"), D("30.25"), D("34.5"), D("41.75")])\n\n\nfrom fractions import Fraction as F\npvariance([F(1, 4), F(5, 4), F(1, 2)])\n\nfrom decimal import Decimal as D\nvariance([D("27.5"), D("30.25"), D("30.25"), D("34.5"), D("41.75")])\n\n\nfrom fractions import Fraction as F\nvariance([F(1, 6), F(1, 2), F(5, 3)])\n\n\ndata = [105, 129, 87, 86, 111, 111, 89, 81, 108, 92, 110,\n        100, 75, 105, 103, 109, 76, 119, 99, 91, 103, 129,\n        106, 101, 84, 111, 74, 87, 86, 103, 103, 106, 86,\n        111, 75, 87, 102, 121, 111, 88, 89, 101, 106, 95,\n        103, 107, 101, 81, 109, 104]\n[round(q, 1) for q in quantiles(data, n=10)]\n\nx = [1, 2, 3, 4, 5, 6, 7, 8, 9]\ny = [1, 2, 3, 1, 2, 3, 1, 2, 3]\ncovariance(x, y)\n\nz = [9, 8, 7, 6,

# **What is the math module?**
It is the C standard defined math module. The math module provides access to mathematical functions.

I know that I have used some jargon in the above statement. Let me clear the clutter. The important terms in the above statement are module, C standard, mathematical functions.

Module: A file containing a set of statements and definitions is called a module. For example, the file program.py is called a module and the module name would be program. You will know how to use a module in the later tutorial.

In [7]:
import math

a = math.sin(math.pi / 2)
print(a)

b = math.factorial(5)
print(b)

c = math.floor(12.6567)
print(c)

d = math.ceil(12.4567)
print(d)

e = math.exp(5)
print(e)

f = math.log(5)
print(f)

g = math.log10(5)
print(g)

h = math.pow(2,6)
print(h)

i = math.sqrt(625)
print(i)

#j = math.abs(-12)
#print(j)


import math

p = 15
q = 20

gcd = math.gcd(p, q)
print(gcd)

import math

n = 5
r = 2

num_permutations = math.perm(n, r)
print(num_permutations)

import math

print(math.factorial(5))

import math

print(math.sqrt(25))  ## Output: 5.0
print(math.sqrt(9))   ## Output: 3.0

import math

print(math.pow(2, 3))  ## Output: 8.0
print(math.pow(5, 2))  ## Output: 25.0

import math

#print(math.abs(-10))  ## Output: 10
#print(math.abs(7.5))  ## Output: 7.5

import math

print(math.sin(math.pi/2))  ## Output: 1.0
print(math.cos(math.pi))    ## Output: -1.0
print(math.tan(math.pi/4))  ## Output: 0.9999999999999999

import math

print(math.asin(1))    ## Output: 1.5707963267948966
print(math.acos(0))    ## Output: 1.5707963267948966
print(math.atan(1))    ## Output: 0.7853981633974483

import math

print(math.log(math.e))   ## Output: 1.0
print(math.log10(100))    ## Output: 2.0

import math

print(math.ceil(3.2))   ## Output: 4
print(math.floor(3.2))  ## Output: 3
print(math.trunc(3.2))  ## Output: 3

import math

print(math.pi)  ## Output: 3.141592653589793
print(math.e)   ## Output: 2.718281828459045
import math

radius = 5
area = math.pi * math.pow(radius, 2)
print(f"The area of the circle with radius {radius} is {area:.2f} square units.")

import math

angle_degrees = 45
angle_radians = math.radians(angle_degrees)

sine = math.sin(angle_radians)
cosine = math.cos(angle_radians)

print(f"The sine of {angle_degrees} degrees is {sine:.2f}")
print(f"The cosine of {angle_degrees} degrees is {cosine:.2f}")

import math

number = 100
natural_log = math.log(number)
base10_log = math.log10(number)

print(f"The natural logarithm of {number} is {natural_log:.2f}")
print(f"The base-10 logarithm of {number} is {base10_log:.2f}")

import math

number = 3.7

ceil_result = math.ceil(number)
floor_result = math.floor(number)
trunc_result = math.trunc(number)

print(f"Ceiling of {number} is {ceil_result}")
print(f"Floor of {number} is {floor_result}")
print(f"Truncated value of {number} is {trunc_result}")

import math

n=int(input('Enter an integer number '))
x=math.fabs(n)
print('Absolute value of %d is %d' %(n,x))

import math

n=int(input('Enter an integer number '))
x=math.sqrt(n)
print('Square root of %d is %d' %(n,x))

import math

n=float(input('Enter a floating point (decimal) number '))
x=math.ceil(n)
print('Ceil value of %f is %f' %(n,x))

import math

n=float(input('Enter a floating point (decimal) number '))
x=math.floor(n)
print('Floor value of %f is %f' %(n,x))

import math

print('Enter two floating point (decimal) numbers')
x=float(input())
y=float(input())
z=math.fmod(x,y)
print('Remainder = %f' %(z))

import math

print('Enter two integer numbers')
a=int(input())
b=int(input())
c=math.pow(a,b)
print('Power = %d' %(c))

import math

x = 10.346
print("ceil", math.ceil(x))
print("floor", math.floor(x))

print("factorial", math.factorial(9))
print("absolute value: ", math.fabs(-10))

print("exponenetial value", math.exp(9))
print("power ", math.pow(2, 3))

angle = math.pi / 4
print("sine", math.sin(angle))
print("cosine", math.cos(angle))
print("tangent", math.tan(angle))

functions = dir(math)


import math
for method in dir(math):
    print(method)




1.0
120
12
13
148.4131591025766
1.6094379124341003
0.6989700043360189
64.0
25.0
5
20
120
5.0
3.0
8.0
25.0
1.0
-1.0
0.9999999999999999
1.5707963267948966
1.5707963267948966
0.7853981633974483
1.0
2.0
4
3
3
3.141592653589793
2.718281828459045
The area of the circle with radius 5 is 78.54 square units.
The sine of 45 degrees is 0.71
The cosine of 45 degrees is 0.71
The natural logarithm of 100 is 4.61
The base-10 logarithm of 100 is 2.00
Ceiling of 3.7 is 4
Floor of 3.7 is 3
Truncated value of 3.7 is 3
Enter an integer number 2
Absolute value of 2 is 2
Enter an integer number 3
Square root of 3 is 1
Enter a floating point (decimal) number 4
Ceil value of 4.000000 is 4.000000
Enter a floating point (decimal) number 3
Floor value of 3.000000 is 3.000000
Enter two floating point (decimal) numbers
4
5
Remainder = 4.000000
Enter two integer numbers
4
5
Power = 1024
ceil 11
floor 10
factorial 362880
absolute value:  10.0
exponenetial value 8103.083927575384
power  8.0
sine 0.7071067811865475
co

# **The Integer, Floating-Point, and String Data Types**
Remember that expressions are just values combined with operators, and they always evaluate to a single value. A data type is a category for values, and every value belongs to exactly one data type. The most common data types in Python are listed in Table 1-2. The values -2 and 30, for example, are said to be integer values. The integer (or int) data type indicates values that are whole numbers. Numbers with a decimal point, such as 3.14, are called floating-point numbers (or floats). Note that even though the value 42 is an integer, the value 42.0 would be a floating-point number. Programmers often use number to refer to ints and floats collectively, although number itself is not a Python data type.

One subtle detail about Python is that any math performed using an int and a float results in a float, not an int. While 3 + 4 evaluates to the integer 7, the expression 3 + 4.0 evaluates to the floating-point number 7.0. Any division between two integers with the / division operator results in a float as well. For example, 16 / 4 evaluates to 4.0 and not 4. Most of the time, this information doesn’t matter for your program, but knowing it will explain why your numbers may suddenly gain a decimal point.